src/transport/ExchangeMgr.h

```diff
@@ -0,0 +1,367 @@
+        /**< Maximum length of string (including NUL character)*/
+    };
+
+    void GetPeerDescription(char * buf, uint32_t bufSize) const;
+
+private:
+    System::PacketBuffer * msg; // If we are re-transmitting, then this is the pointer to the message being retransmitted
+
+    // Trickle-controlled retransmissions:
+    uint32_t backoff;             // backoff for sampling the numner of messages
+    uint8_t msgsReceived;         // number of messages heard during the backoff period
+    uint8_t rebroadcastThreshold; // re-broadcast threshold
+    uint16_t mFlags;              // Internal state flags
+
+    CHIP_ERROR ResendMessage();
+    bool MatchExchange(PeerConnectionState * msgCon, const PacketHeader & packetHeader, const ChipExchangeHeader * exchangeHeader);
+    static void CancelRetransmissionTimer(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+    static void TimerTau(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+    static void TimerT(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+
+    CHIP_ERROR StartResponseTimer();
+    void CancelResponseTimer();
+    static void HandleResponseTimeout(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+
+    uint32_t mPendingPeerAckId;
+    void DoClose(bool clearRetransTable);
+    CHIP_ERROR HandleMessage(const PacketHeader & packetHeader, const ChipExchangeHeader * exchHeader,
+                             System::PacketBuffer * msgBuf);
+    CHIP_ERROR HandleMessage(const PacketHeader & packetHeader, const ChipExchangeHeader * exchHeader,
+                             System::PacketBuffer * msgBuf, ExchangeContext::MessageReceiveFunct umhandler);
+    void HandleConnectionClosed(CHIP_ERROR conErr);
+
+    uint8_t mRefCount;
+};
+
+/**
+ *  @class ExchangeManager
+ *
+ *  @brief
+ *    This class is used to manage ExchangeContexts with other CHIP nodes.
+ *    It works on behalf of higher layers, creating ExchangeContexts and
+ *    handling the registration/unregistration of unsolicited message handlers.
+ *
+ */
+class DLL_EXPORT ExchangeManager
+{
+    friend class ExchangeContext;
+
+public:
+    enum State
+    {
+        kState_NotInitialized = 0, /**< Used to indicate that the ExchangeManager is not initialized */
+        kState_Initialized    = 1  /**< Used to indicate that the ExchangeManager is initialized */
+    };
+
+    ExchangeManager();
+    ExchangeManager(const ExchangeManager &) = delete;
+    ExchangeManager operator=(const ExchangeManager &) = delete;
+
+    SecureSessionMgrBase * MessageLayer; /**< [READ ONLY] The associated SecureSessionMgrBase object. */
+    uint8_t State;                       /**< [READ ONLY] The state of the ExchangeManager object. */
+
+    CHIP_ERROR Init(SecureSessionMgrBase * msgLayer);
+    CHIP_ERROR Shutdown();
+
+    ExchangeContext * NewContext(const uint64_t & peerNodeId, void * appState = nullptr);
```

Comment:
These will need to be updated to use PeerAddress and transports.

Reply:
ExchangeMgr is on top of SecureSessionMgr instead of Transport. SecureSessionMgr only support to specify the node ID as the destination.